### ISSUE TITLE
Race condition in OIDC auth in browser

### ISSUE THREAD

**Issue by g7r** [hashicorp/vault#26103]
**Describe the bug**
There is a really nasty issue in OIDC authentication that bugged me for quite some time. Sometimes I had to retry "Sign in with OIDC provider" like tens of times to succeed. The problem reproduced only in Firefox and only on my laptop. However, the experience was maddening so I had to figure it out.

It seems like there is a race between the window with button "Sign in with OIDC provider" and the popup window, that does second part of authentication.

**To Reproduce**
1. Use Firefox. Set `dom.min_background_timeout_value_without_budget_throttling` to 100000 ms in about:config to make the issue 100% reproducible (default value is 1000 ms).
2. Open sign in page (Vault should be set up with OIDC auth)
3. Press "Sign in with OIDC provider"
4. Wait a little bit
5. See the error "The provider window was closed before authentication was complete. Please click Sign In to try again".

On my particular hardware+software setup on the default Firefox settings the reproducibility is in the ballpark of 90%.

**Expected behavior**
Sign in should succeed.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.12.2
* Vault CLI Version (retrieve with `vault version`): Vault v1.15.6 (cgo) (not needed to reproduce the issue)
* Server Operating System/Architecture: Manjaro Linux 23.1.3
* Firefox: 124.0

**Additional context**
I believe there is some kind of sleep in JavaScript in the window with the "Sign in with OIDC provider" button. This sleep is being throttled when the window goes to background. This results in waking up too late, when the popup window is already closed.

If I manually switch focus from the popup window to the main one, making it foreground, the issue doesn't reproduce.

**Comment by hsimon-hashicorp** [comment]
Hi! Are you able to reproduce this on current versions of Vault? For bug reports, we have historically been accepting reproducible issues for versions n-2. Thanks!

**Comment by g7r** [comment]
I can't force corporate Vault server upgrade unfortunately. To reproduce the issue on a more recent Vault version I need to set it up myself. Maybe there is a manual somewhere on setting up a minimal local Vault server with OIDC. I'm not a Vault operations expert.

**Comment by hashishaw** [comment]
Hi @g7r, thanks for the report. Unfortunately I've been unable to reproduce on our n-2 versions (back to 1.14.x). I'm interested if you'd be able to get it to fail on more recent versions, as we have seen some linux-specific bugs in the past. 

We have some tutorials here: https://developer.hashicorp.com/vault/docs/auth/jwt/oidc-providers.

Here's a simple script that our team uses for setting up OIDC with Auth0 locally (I have a dev account that I use, it's free and fairly easy to set up): 
```
# set these from your auth0 account
AUTH0_DOMAIN=
AUTH0_CLIENT_ID=
AUTH0_CLIENT_SECRET=

# VARIABLES: 
mount_path=oidc # update if you want to configure with a custom path (/oidc is the default)

vault auth enable -path=$mount_path oidc

vault write auth/$mount_path/config oidc_discovery_url="https://$AUTH0_DOMAIN/" oidc_client_id="$AUTH0_CLIENT_ID" oidc_client_secret="$AUTH0_CLIENT_SECRET" default_role="reader"

vault write auth/$mount_path/role/reader \
    bound_audiences="$AUTH0_CLIENT_ID" \
    allowed_redirect_uris="http://localhost:8250/oidc/callback" \
    allowed_redirect_uris="http://localhost:8200/ui/vault/auth/$mount_path/oidc/callback" \
    user_claim="sub" \
    policies="reader"
```
It may also be interesting to check this setup on your current version, as maybe there is a specific OIDC config that is causing this. 

Hope that helps, and please let us know what you find!

**Comment by g7r** [comment]
Hi @hashishaw! I've tried to reproduce the issue on local Vault cluster using current version and didn't succeed. But I didn't try my repro on 1.12.2 (the version that we currently use). So I don't know for sure whether I missed something in my test setup or the problem was fixed in recent Vault versions.

I've solved the problem for myself by adjusting `dom.min_background_timeout_value_without_budget_throttling` to 100 ms, so I have little incentive to invest more time into the issue. I think that means that we should close my issue. If I encounter this problem on fresher Vault versions I'll come with another bug report.

For now my report could be useful for those who suffer from the same problem if there are any :)
